client/TracySysTrace.cpp

```diff
@@ -796,11 +797,180 @@ static void SetupSampling( int64_t& samplingPeriod )
+//    Only command-line programs can be run as root. This means that we can't
+//    limit support to root processes, we have to support unprivileged user processes.
+//    As some of the things that we need to do require root privileges, this forces
+//    us to spawn sub-processes executing commands as root. On rooted Android devices,
+//    the default `su` command succeeds non-interactively (does not ask for any
+//    credentials) so this is mostly transparent to the user.
+// 2. On Android, multiple implementations of `su` are in use, with different
+//    command-line syntax and semantics. The default `su` in Android has the following
+//    syntax to run a command as root:
+//      su root some_command
+//    Moreover, it directly exec's the command, there is no
+//    shell invocation, which means that commands that need to be interpreted by a
+//    shell (e.g. commands using the > redirection operator) need to be wrapped
+//    in a `sh -c` invokation, like this:
+//      su root sh -c 'some_command'
+//    This is in contrast with other `su` commands, which typically have syntax
+//      su -c some_command
+//    and which tend to execute a shell to interprete the command (so there is
+//    no need for `sh -c`). Some developers' Android device have such `su` commands,
+//    for instance from https://github.com/topjohnwu/Magisk.
+//    We have to handle both flavors of `su` commands.
+
+// Internal implementation helper for GetHowToRunAsRoot.
+//
+// Checks if `su <flag> command` succeeds running the specified command as root.
+//
+// Concretely, the test performed is to run the `id -u` command as root, and check
+// that its output is `0` (the uid of root).
+// This test is based on the command's stdout and not on the return code, because
+// it seems that some implementations of `su` don't propagate the command's return
+// code.
+static bool TrySuCommandFlag(const char* flag) {
+    bool success = false;
+    int pipefd[2];
+    if( pipe( pipefd ) == 0 )
+    {
+        int read_end = pipefd[0];
+        int write_end = pipefd[1];
+        const auto pid = fork();
+        if( pid == 0 )
+        {
+            // child
+            close( read_end );
+            dup2( write_end, STDOUT_FILENO );
+            close( write_end );
+            execlp( "su", "su", flag, "id", "-u", nullptr );
+            exit( EXIT_FAILURE );
+        }
+        else if( pid > 0 )
+        {
+            // parent
+            close( write_end );
+            char buf[8] = { 0 };
+            int read_len = read( read_end, buf, sizeof buf );
+            if (read_len >= 2) {
+                success = !memcmp(buf, "0\n", 2 );
+            }
+            close( read_end );
+            waitpid( pid, nullptr, 0 );
+        }
+    }
+    return success;
+}
+
+// Enum identifying a method for running a command as root.
+enum class HowToRunAsRoot {
+    // Our process is already root (getuid()==0). Nothing else is needed.
+    AlreadyRunningAsRoot,
```

Comment:
Is there a scenario in which this case would be possible?

Reply:
Expanded the comment to explain that scenario.